File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
using System.Collections.Generic;

public static class ProteinTranslation
{
    private const int CodonLength = 3;

    public static string[] Proteins(string strand)
    {
        var proteins = new List<string>();
        var numberOfCodons = strand.Length / CodonLength;

        for (var i = 0; i < numberOfCodons; i++)
        {
            if (strand.Length - i * CodonLength < CodonLength)
                break; // Not enough nucleotides left to form a full codon

            var codon = strand.Substring(i * CodonLength, CodonLength);
            var proteinName = ProteinNames.FromCodon(codon);

            if (proteinName == ProteinNames.Stop)
                break;

            proteins.Add(proteinName);
        }

        return proteins.ToArray();
    }
}

public static class ProteinNames
{
    public static string Methionine = "Methionine";
    public static string Phenylalanine = "Phenylalanine";
    public static string Leucine = "Leucine";
    public static string Serine = "Serine";
    public static string Tyrosine = "Tyrosine";
    public static string Cysteine = "Cysteine";
    public static string Tryptophan = "Tryptophan";
    public static string Stop = "STOP";

    public static string FromCodon(string codon)
        => Codons.TryGetValue(codon, out var proteinName)
            ? proteinName
            : Stop;

    public static Dictionary<string, string> Codons = new()
    {
        { "AUG", Methionine },
        { "UUU", Phenylalanine },
        { "UUC", Phenylalanine },
        { "UUA", Leucine },
        { "UUG", Leucine },
        { "UCU", Serine },
        { "UCC", Serine },
        { "UCA", Serine },
        { "UCG", Serine },
        { "UAU", Tyrosine  },
        { "UAC", Tyrosine },
        { "UGU", Cysteine },
        { "UGC", Cysteine },
        { "UGG", Tryptophan },
        { "UAA", Stop },
        { "UAG", Stop },
        { "UGA", Stop }
    };
}

File: solutions/ruby/protein-translation/1/protein_translation.rb
class Translation
  def self.of_rna(strand)
    strand
      .split(/(.{3})/)
      .each_with_object([]) do |c, p|
        next if c.empty?
        protein = Protein.from_codon(c)
        break p if protein == "STOP"
        p << protein
      end
  end
end

class Protein
  CODONS = {
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UGG" => "Tryptophan",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP"
  }.freeze

  def self.from_codon(codon)
    raise InvalidCodonError.new(codon) unless CODONS.key?(codon)

    CODONS.fetch(codon)
  end
end

class InvalidCodonError < StandardError
  def initialize(codon)
    super("Invalid codon: #{codon}.")
  end
end
